lkl tools: introduce Intel DPDK network backend for virtio device

thehajime opened the pull request:
This involves several refactors in lkl tools:
- split tap code from hijack library to general library
- change env name for interface name of tap
  old) LKL_HIJACK_NET_TAP
  new) LKL_HIJACK_NET_IFTYPE and LKL_HIJACK_NET_IFNAME
- use opqaue pointers for lkl_netdev.
  Using opaque pointers remove the implementation details from lkl.h
  and increase the scalability for additional device emulation backends.
- expose common test functions and macros into tests/test.h

only tested with dpdk-2.2.0, vmxnet3 driver. at least e1000 and e1000e
don't work on vmware fusion (6.0.x).

to build with dpdk backend, one needs to do in advance:
- ./scripts/dpdk-sdk-build.sh
- make dpdk=yes

Signed-off-by: Thomas Liebetraut thomas@tommie-lie.de
Signed-off-by: Hajime Tazaki thehajime@gmail.com

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/lkl/linux/67)

<!-- Reviewable:end -->

thehajime commented:
the way to specify to build in DPDK is an open question.
there are a couple of choices: currently the former was implemented.

- ```make``` argument (make DPDK=yes)
- CONFIG_LKL_DPDK like option ? (configured by make xxxconfig ARCH=lkl)

thehajime commented:
thanks @stfairy for the comment. I'll get you back addressing your comments (I think I'm fine:)).

thehajime commented:
rebased and split into commits.

for additional information how to use dpdk, I drafted a wiki page, which could be put in lkl/linux later.

https://github.com/libos-nuse/lkl-linux/wiki/DPDK-Howto-(tmp)

pscollins commented:
Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 64 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQCRADPfNacAHEA5P:-KBAQCRADPfNacAHEA5Q:1185478267) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L64)):*
Is this threadsafe? Does it require any synchronization? 

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 83 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQIVMI-V44CItV651:-KBAQIVMI-V44CItV652:957668506) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L83)):*
Ditto --- threadsafe?

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 97 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAPHSvFXFOaJbPzeru:-KBAPHSvFXFOaJbPzerv:-190004938) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L97)):*
What's this sleep for?

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 103 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAPXbYaWgsn5xXolB4:-KBAPXbYaWgsn5xXolB5:-256789823) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L103)):*
No need for `rm0` here, just assign straight to `rm`

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 117 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAPq1uMe4mdzWzoKqz:-KBAPq1vH0Z7FpgRBDwa:711689482) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L117)):*
I imagine this is a fairly regular occurrence, you probably want to wrap it in an `#ifdef DEBUG` (or define a `fprintf` equivalent that's wrapped like that)

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 149 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAOfC3AfVN6tDb6d-k:-KBAOfC3AfVN6tDb6d-l:109621491) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L149)):*
Why doesn't this call `poll(2)`? `net_enqueue` in `tools/lkl/lib/virtio_net.c` expects this to be reasonably honest, because it doesn't want to block on a read/write call and hold on to a lock for longer than needed. 

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 196 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAMHSv2QA2RbHeyq2q:-KBAMHSv2QA2RbHeyq2r:1231699485) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L196)):*
I think you should use NULL if you're returning a pointer; it's more intention-revealing

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 271 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBALRY-rAEBKgJQpxev:-KBALRY-rAEBKgJQpxew:-763966328) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L271)):*
IMO this should be a compile-time error, with the #error pragma

---

*[tools/lkl/lib/virtio_net_tap.c, line 5 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAKFeax177EMDbYNAT:-KBAKFebnWmUodsJkcp8:2118113474) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_tap.c#L5)):*
You might want to add at @opurdila here

---

*[tools/lkl/lib/virtio_net_tap.c, line 24 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAKHNqbSVgAxey7Uuk:-KBAKHNqbSVgAxey7Uul:-692703816) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_tap.c#L24)):*
Upstream has changed so these aren't needed anymore, you might want to rebase on to it

---

*[tools/lkl/lib/virtio_net_tap.c, line 40 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAJChCPwPNOIPS_nW4:-KBAJChCPwPNOIPS_nW5:704223687) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_tap.c#L40)):*
I'm not sure if it's easily available here, but the container_of macro in .e.g include/lkl/linux/virtio_net.h is a nice way to do this that allows for struct members to be reordered down the line.

---

*[tools/lkl/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBALhywP_UkWkv5WIB1:-KBALhywP_UkWkv5WIB2:1670837715) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/Makefile#L47)):*
IMO this should be something like `CONFIG_LKL_DPDK`

---

*[tools/lkl/scripts/dpdk-sdk-build.sh, line 5 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBALnGdcS85ZhsSPacr:-KBALnGdcS85ZhsSPacs:1038982021) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/scripts/dpdk-sdk-build.sh#L5)):*
I don't think you should clone git repos as part of the build process --- maybe add it as a submodule? Someone who knows their build processes better than I do should weigh in --- @tommie-lie?

---

*[tools/lkl/tests/net-test.c, line 112 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQyo7ElFD-UXFsA7x:-KBAQyo7ElFD-UXFsA7y:-327695201) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net-test.c#L112)):*
It's not safe to do this until after you've checked `argc`

---

*[tools/lkl/tests/net-test.c, line 195 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAR4bwLrm4JnU-1SPn:-KBAR4bwLrm4JnU-1SPo:-2011683416) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net-test.c#L195)):*
`endif /* __MIGW32__ */`

---

*[tools/lkl/tests/net-test.c, line 204 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAR9Fb__cAfAM7hjGz:-KBAR9Fb__cAfAM7hjH-:-1168765263) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net-test.c#L204)):*
`endif /* __MIGW32__ */`

Though it looks like you might not want this file to be compiled at all if we're on Windows?

---

*[tools/lkl/tests/net.sh, line 11 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBANJ2A3bJFkcncJ_b9:-KBANJ2A3bJFkcncJ_bA:-1989586520) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net.sh#L11)):*
Where are the DPDK tests run? Are they hard to do in a platform-independent way?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/67)*
<!-- Sent from Reviewable.io -->

thehajime commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 64 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQCRADPfNacAHEA5P:-KBHrrWN5sf0rZC8sjWY:422409175) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L64)):*
good catch. some reported DPDK's mempool with cache is not thread safe (e.g., http://www.dpdk.io/ml/archives/dev/2014-February/001401.html). not sure if it's already fixed or not. so I will disable the cache by changing MEMPOOL_CACHE_SZ to 0.

other than that, I don't see any worries on thread safeness.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 83 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQIVMI-V44CItV651:-KBI0akThRgwczxiEi4H:-624460677) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L83)):*
```nd_dpdk->rms``` is potentially not safe in multiple threads.

if ```virtio_process_one()``` is called by multiple threads then this should be guarded.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 97 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAPHSvFXFOaJbPzeru:-KBHvMlM_C6-7ANdAicX:709922801) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L97)):*
see my another comments on poll(2) below.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 103 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAPXbYaWgsn5xXolB4:-KBHtzavBYw9rnJj24er:1388568597) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L103)):*
okay. will fix it.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 117 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAPq1uMe4mdzWzoKqz:-KBHu9IbQshnyZLWx1NL:1912028314) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L117)):*
okay. will wrap with ifdef DEBUG.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 149 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAOfC3AfVN6tDb6d-k:-KBHujUUCfRAmd6xCq7U:-585086031) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L149)):*
dpdk's interrupt mode has equivalent of epoll_wait(2), which we can apply here. but AFAIK the mode is only available on limited NIC drivers like ixgbe/igb/e1000 (with dpdk v2.2.0), while my test environment uses vmxnet3.

```sleep()``` after ```rte_eth_rx_burst``` needs for this reason to relax processor.

so overall of this patch, I tried to be available as much drivers as possible.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 196 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAMHSv2QA2RbHeyq2q:-KBHwF0OA8kwPGh2lBQM:-1227134214) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L196)):*
thanks. will fix those.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 271 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBALRY-rAEBKgJQpxev:-KBHwLif9YQbgG1u8gn3:-408304170) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L271)):*
since this function is called via library (sometime via LD_PRELOAD with configuration of env variable), giving helpful message to a library user isn't harmful IMHO. 

e.g.,
```
sudo LKL_HIJACK_NET_VIF=dpdk LKL_HIJACK_NET_IFNAME=dpdk LKL_HIJACK_NET_IP=192.168.13.2 LKL_HIJACK_NET_NETMASK_LEN=24  LKL_HIJACK_DEBUG=1 ./bin/lkl-hijack.sh ip ad
```

since we/I don't have any cool manager of each VIF right now, doing it in each VIF is the way to be polite. what do you think ?

---

*[tools/lkl/lib/virtio_net_tap.c, line 5 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAKFeax177EMDbYNAT:-KBHyRIW_t_iVhS9UGeK:-353913272) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_tap.c#L5)):*
indeed, thanks !

---

*[tools/lkl/lib/virtio_net_tap.c, line 24 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAKHNqbSVgAxey7Uuk:-KBHycrnS4Yf78FKlVQu:1099770996) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_tap.c#L24)):*
will fix it. thanks.

---

*[tools/lkl/lib/virtio_net_tap.c, line 40 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAJChCPwPNOIPS_nW4:-KBHyu2nZqxq7JJ1tJQL:1119235220) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_tap.c#L40)):*
let me think a bit more for this: I will keep this opened.
(though I agree with you for the usefulness of container_of macro)

---

*[tools/lkl/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBALhywP_UkWkv5WIB1:-KBHz5L5hS0AQv1bD2iK:-1311219006) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/Makefile#L47)):*
will see how #82 reaches a consensus. 

---

*[tools/lkl/scripts/dpdk-sdk-build.sh, line 5 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBALnGdcS85ZhsSPacr:-KBHzHPqrjw4yqmwVBIu:-327048981) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/scripts/dpdk-sdk-build.sh#L5)):*
this script is never called from build system:  it is intended by people by hand in advance, with the help of wiki document.

and I think having git submodule in linux tree is hard to convince again so, I just crafted this script for a work around.

---

*[tools/lkl/tests/net-test.c, line 112 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQyo7ElFD-UXFsA7x:-KBHzhWJlIU6Nyy8OgVC:1388568597) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net-test.c#L112)):*
okay. will fix it.

---

*[tools/lkl/tests/net-test.c, line 204 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAR9Fb__cAfAM7hjGz:-KBI-HYwgU_wS0-sGU7y:1761210376) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net-test.c#L204)):*
right, as lib/virtio_net* are filtered out in ```tools/lkl/Makefile``` on mingw32 build.

---

*[tools/lkl/tests/net.sh, line 11 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBANJ2A3bJFkcncJ_b9:-KBI-buEd6c3GNXnWlV5:-2123969700) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net.sh#L11)):*
no idea for non-linux platform at this moment.
and preparations to use dpdk are complicated as documented in wiki.

so I don't see any benefit to do a test suite.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/67)*
<!-- Sent from Reviewable.io -->

pscollins commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 64 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQCRADPfNacAHEA5P:-KBJSxEjCm10ccJiYH2b:282946994) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L64)):*
Okay --- it might be nice to stick in a comment explaining when it could be dangerous to use (in the TAP case, it's easy to tell that read(2) or write(2) are okay, but it's less obvious here).

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 83 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQIVMI-V44CItV651:-KBJTNbQ2JbOzLh3GiM6:1923379884) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L83)):*
Ditto, please stick in a comment so that e.g. after a future refactor that allows parallel writes, if something breaks it will be easier to figure out where to start. Also, virtio_process_one can be called by 2 threads concurrently (1 per queue), so be aware of that.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 149 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAOfC3AfVN6tDb6d-k:-KBJTuy0Aaq0Fv34Hu7W:-1518375482) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L149)):*
Hmmm, okay, that's unfortunate. Oh well, nothing to be done about it I guess.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 271 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBALRY-rAEBKgJQpxev:-KBJU8K9767-9fzm_xNC:-1793766723) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L271)):*
Makes sense. In my head I imagined there was a way to make it so this symbol isn't available unless LKL was compiled with DPDK, but I guess that's not the case (for now, at least).

---

*[tools/lkl/tests/net.sh, line 11 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBANJ2A3bJFkcncJ_b9:-KBJUsdf34ZE1uH_fv6Y:1927902097) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net.sh#L11)):*
Well, I guess what I really meant to say was "Don't we want this running on circleCI?" because I worry that it will get broken and no one will notice until after the fact. If you don't think it's worth it, then no worries.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/67)*
<!-- Sent from Reviewable.io -->

stfairy commented:
Sorry I didn't notice there were some comments on Reviewable so maybe some of my comments were duplicate.

Aside from that, can you also split this pull request into two, so your first refactoring commit can get merged early before people can fully review the DPDK part?

The refactoring looks good to me except for a few minor details.

Thanks,
Xiao

thehajime commented:
> Aside from that, can you also split this pull request into two, so your first refactoring commit can get merged early before people can fully review the DPDK part?

agree: will keep dpdk part to this PR and split refactor part to new PR.

thehajime commented:
I split tap refactor part to another PR (#118): please comment the tap part to the PR if you have, not in this PR.

thehajime commented:
Reviewed 1 of 7 files at r2.
Review status: 3 of 20 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 64 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQCRADPfNacAHEA5P:-KCNV0MsATSuOX3VW461:-513363225) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L64)):*
let me know if the current commit addresses your request.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 83 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQIVMI-V44CItV651:-KCNW2v0rkeYijoeXJHN:-1618680090) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L83)):*
this too: let me how you feel with new comment.

---

*[tools/lkl/tests/net-test.c, line 195 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAR4bwLrm4JnU-1SPn:-KCNWWn-iy5gAiZl8uQP:-1768833933) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net-test.c#L195)):*
added.

---

*[tools/lkl/tests/net.sh, line 11 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBANJ2A3bJFkcncJ_b9:-KCNYEyjo_hTXzZ7pULZ:-1988805372) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/tests/net.sh#L11)):*
I added a dummy test in tests/net.sh for DPDK so that users can customize the test. it is not likely to do a generic test in all environment. if someone comes up with a nice test for DPDK, it would be great but I have no idea.

on CircleCI, neither tap nor DPDK are available. more likely way would be use of raw socket with a root privilege.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/67)*
<!-- Sent from Reviewable.io -->

pscollins commented:
Reviewed 1 of 17 files at r3, 1 of 5 files at r4.
Review status: 4 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 64 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQCRADPfNacAHEA5P:-KCWsy_EWbgPKyzNWeIu:1794731161) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L64)):*
Great!

---

*[tools/lkl/lib/virtio_net_dpdk.c, line 83 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/67#-KBAQIVMI-V44CItV651:-KCWszfvj6VYzJ7nXOuZ:-1611012647) ([raw file](https://github.com/lkl/linux/blob/b950200f8895692180dfbbbe21c9e0d12e9fcfdd/tools/lkl/lib/virtio_net_dpdk.c#L83)):*
Perfect!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/67)*
<!-- Sent from Reviewable.io -->

thehajime commented:
I modified a bit (the name of ifdefs) and rebased it.

thehajime commented:
I created a wiki page (https://github.com/lkl/linux/wiki/Howto:-DPDK-with-LKL) for the DPDK introduction.
